This folder contains files with the expression data. The columns of the files are semicolon-separated and contain headers.

### Sample description

| <sub> File name  </sub>                                       | <sub> Species </sub> | <sub> Expression estimation </sub> | <sub> Normalization </sub> | <sub> Metric </sub>    |
| :------------------------------------------------------------ | :------------------- | :--------------------------------- | :------------------------- | :--------------------- |
| <sub> AL_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> AL </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> AL_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> AL </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> AL_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> AL </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> AL_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> AL </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> AT_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> AT </sub>| <sub> Gene-level </sub>      | <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> AT_genes_inter_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> AT </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> VST counts </sub>| 
| <sub> AT_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> AT </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> AT_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> AT </sub>|<sub> Transcript-level </sub>|<sub> DESeq intra-organ</sub>|<sub> VST counts</sub>| 
| <sub> AT_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> AT </sub>|<sub> Transcript-level </sub> | <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> AT_transcripts_inter_norm_count_mat_vsd_sample_names.csv </sub>       | <sub> AT </sub>|<sub> Transcript-level </sub>|<sub> DESeq inter-organ</sub>|<sub> VST counts</sub>| 
| <sub> AT_transcripts_intra_norm_count_mat_vsd_sample_names.csv </sub>       | <sub> AT </sub>|<sub> Transcript-level </sub>|<sub> DESeq intra-organ</sub>|<sub> VST counts</sub>| 
| <sub> BD_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> BD </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> BD_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> BD </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> BD_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> BD </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> BD_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> BD </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> CR_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> CR </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> CR_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> CR </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> CR_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> CR </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> CR_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> CR </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> ES_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> ES </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> ES_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> ES </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> ES_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> ES </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> ES_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> ES </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> MT_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> MT </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> MT_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> MT </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> MT_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> MT </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> MT_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> MT </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> TH_genes_complete_table_tpm_with_circRNA_sample_names.csv </sub>      | <sub> TH </sub>| <sub> Gene-level </sub> | <sub> none </sub>             | <sub> TPM </sub>       | 
| <sub> TH_genes_inter_norm_tpm_mat_deseq_sample_names.csv </sub>             | <sub> TH </sub>| <sub> Gene-level </sub> | <sub> DESeq inter-organ </sub>| <sub> TPM </sub>       | 
| <sub> TH_genes_intra_norm_count_mat_vsd_sample_names.csv </sub>             | <sub> TH </sub>| <sub> Gene-level </sub> | <sub> DESeq intra-organ </sub>| <sub> VST counts </sub>| 
| <sub> TH_transcripts_complete_table_tpm_with_circRNA_sample_names.csv </sub>| <sub> TH </sub>| <sub> Transcript-level </sub>| <sub> none </sub>        | <sub> TPM </sub>       | 
| <sub> AT_Pt_Mt_Orthologs.csv </sub>                                         | <sub> AT </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> AL_Pt_Mt_Orthologs.csv </sub>                                         | <sub> AL </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> CR_Pt_Mt_Orthologs.csv </sub>                                         | <sub> CR </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> ES_Pt_Mt_Orthologs.csv </sub>                                         | <sub> ES </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> TH_Pt_Mt_Orthologs.csv </sub>                                         | <sub> TH </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> MT_Pt_Mt_Orthologs.csv </sub>                                         | <sub> MT </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 
| <sub> BD_Pt_Mt_Orthologs.csv </sub>                                         | <sub> BD </sub>| <sub> NA </sub>         | <sub> NA </sub>               | <sub> NA </sub>        | 


<br/>

`inter-organ` indicates that data was normalized between organs, whereas `intra-organ` indicates that data was only normalized within the same organ type (developmental stages and replicates).

Species: `AT` = *Arabidopsis thaliana*; `AL` = *Arabidopsis lyrata*; `CR` = *Capsella rubella*; `ES` = *Eutrema salsugineum*; `TH` = *Tarenaya hassleriana*; `MT` = *Medicago truncatula*; `BD` = *Brachypodium distachyon*; 
